omnigibson/transition_rules.py

```diff
@@ -103,551 +263,952 @@ def __call__(self, obj):
+    the recipe event
+    """
+    @classproperty
+    def candidate_filters(cls):
+        # Modify the container filter to include "blender" ability as well
+        candidate_filters = super().candidate_filters
+        candidate_filters["container"] = AndFilter(filters=[candidate_filters["container"], AbilityFilter(ability="blender")])
+        return candidate_filters
+
+    @classmethod
+    def _generate_conditions(cls):
+        # Container must be toggledOn, and should only be triggered once
+        return [ChangeConditionWrapper(
+            condition=StateCondition(filter_name="container", state=ToggledOn, val=True, op=operator.eq)
+        )]
+
+
+class MixingToolRule(MixingRule):
+    """
+    Transition mixing rule that leverages "mixing_tool" ability objects, which require touching between a mixing tool
+    and a container in order to trigger the recipe event
+    """
+    @classmethod
+    def add_recipe(cls, name, input_objects, input_systems, output_system):
+        # We do not allow any input objects to be specified! Assert empty list
+        assert len(input_objects) == 0, f"No input_objects should be specified for {cls.__name__}!"
+
+        # Call super
+        super().add_recipe(
+            name=name,
+            input_objects=input_objects,
+            input_systems=input_systems,
+            output_system=output_system,
+        )
+
+    @classproperty
+    def candidate_filters(cls):
+        # Add mixing tool filter as well
+        candidate_filters = super().candidate_filters
+        candidate_filters["mixing_tool"] = AbilityFilter(ability="mixing_tool")
+        return candidate_filters
+
+    @classmethod
+    def _generate_conditions(cls):
+        # Mixing tool must be touching the container, and should only be triggered once
+        return [ChangeConditionWrapper(
+            condition=TouchingAnyCondition(filter_1_name="container", filter_2_name="mixing_tool")
+        )]
+
+    @classproperty
+    def ignore_nonrecipe_objects(cls):
+        return True
 
 
-"""See the following example for writing simple rules.
-
-  GenericTransitionRule(
-      filters=(
-          AndFilter(CategoryFilter("light"), StateFilter(ToggledOn, True)),
-          CategoryFilter("key"),
-      ),
-      condition_fn=lambda light, key: dist(light, key) < 1,
-      transition_fn=lambda light, key: light.states[ToggledOn].set_value(True),
-  )
-"""
-# TODO: To prevent bugs for now, we make this READ-ONLY (tuple). In the future, rules should be allowed to be added
-# dynamically
-DEFAULT_RULES = (
-    SlicingRule(),
-    DicingRule(),
-    # Strawberry smoothie
-    BlenderRule(
-        output_system="strawberry_smoothie",
-        particle_requirements={"milk": 10},
-        obj_requirements={"strawberry": 5, "ice_cube": 5},
-    ),
+# Create strawberry smoothie blender rule
+BlenderRule.add_recipe(
+    name="strawberry_smoothie_recipe",
+    input_objects={"strawberry": 5, "ice_cube": 5},
```

Comment:
Are we going to check for "categorical correctness" only? I guess we can still use your code here, but set "strawberry": 1 and "ice_cube": 1.

Reply:
exactly. at least this future-proofs us in case we do implement some more fine-grained, quantity-based requirement